genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobTerminationServiceImpl.java

```diff
@@ -0,0 +1,118 @@
+import com.netflix.genie.common.exceptions.GenieException;
+import com.netflix.genie.common.exceptions.GenieServerException;
+import com.netflix.genie.proto.JobKillListenerRequest;
+import com.netflix.genie.proto.JobKillListenerResponse;
+import com.netflix.genie.proto.JobTerminationServiceGrpc;
+import com.netflix.genie.web.rpc.grpc.interceptors.SimpleLoggingInterceptor;
+import com.netflix.genie.web.services.JobKillServiceV4;
+import com.netflix.genie.web.services.JobSearchService;
+import io.grpc.stub.StreamObserver;
+import lombok.extern.slf4j.Slf4j;
+import net.devh.springboot.autoconfigure.grpc.server.GrpcService;
+
+import javax.validation.constraints.NotBlank;
+import java.util.Map;
+
+/**
+ * Extension of {@link JobTerminationServiceGrpc.JobTerminationServiceImplBase}
+ * to provide functionality for killing jobs running on agents.
+ * TODO Register with HeartBeatService to listen for agent stream to become inactive and clean it up.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@GrpcService(
+    value = JobTerminationServiceGrpc.class,
+    interceptors = {
+        SimpleLoggingInterceptor.class,
+    }
+)
+@Slf4j
+public class GRpcJobTerminationServiceImpl
+    extends JobTerminationServiceGrpc.JobTerminationServiceImplBase
+    implements JobKillServiceV4 {
+
+    private final Map<String, StreamObserver<JobKillListenerResponse>> activeStreams = Maps.newConcurrentMap();
+    private final JobSearchService jobSearchService;
+
+    /**
+     * Constructor.
+     * @param jobSearchService      Job search service
+     */
+    public GRpcJobTerminationServiceImpl(final JobSearchService jobSearchService) {
+        this.jobSearchService = jobSearchService;
+    }
+
+    /**
+     * Register to get a callback when a job is killed.
+     *
+     * @param request          Request to listen for job registerCallback
+     * @param responseObserver The response observer
+     */
+    @Override
+    public void registerForKillNotification(
+        final JobKillListenerRequest request,
+        final StreamObserver<JobKillListenerResponse> responseObserver
+    ) {
+        activeStreams.put(request.getJobId(), responseObserver);
+    }
+
+    @Override
+    public void killJob(final @NotBlank(message = "No id entered. Unable to kill job.") String id,
+                        final @NotBlank(message = "No reason provided.") String reason) throws GenieException {
+
+        final StreamObserver<JobKillListenerResponse> responseObserver = activeStreams.get(id);
+
+        if (!jobSearchService.getJobStatus(id).isFinished()) {
+            if (responseObserver != null) {
+                responseObserver.onNext(
+                    JobKillListenerResponse.newBuilder().setJobId(id).build()
+                );
+                responseObserver.onCompleted();
+
+
+                log.info("Agent communicated for killing the job ");
+            } else { //This is bad. No responseobserver for a unfinished job. Tell the user
```

Comment:
Comment can be improved. But it is bad in the sense that this should not happen. That tells we routed to the wrong box. This will be executed in response to the kill request from a user. Since we are throwing the exception the exception will go to the user. Hence the comment(which I will change) but the user will know